File: java/parsers/src/main/java/jp/co/teruuu/parser/json/JParser.java
package jp.co.teruuu.parser.json;

import jp.co.teruuu.parser.common.Parser;
import jp.co.teruuu.parser.json.type.JValue;

public interface JParser<T> extends Parser<T> {

  default JParser<JValue> or(JParser<JValue> rhs) {
    return new JOrParser(this, rhs);
  }
  static JParser<JValue> bool() {
    return new JBoolParser();
  }
  static JParser<JValue> nu() {
    return new JNullParser();
  }
  static JParser<JValue> number() {
    return new JNumberParser();
  }
  static JParser<JValue> string() {
    return new JStringParser();
  }

  static JParser<JValue> object() {
    return new JObjectParser();
  }
  static JParser<JValue> array() {
    return new JArrayParser();
  }
  static JParser<JValue> value() {
    return JParser.bool().or(JParser.nu()).or(JParser.number()).or(JParser.string()).or(JParser.array()).or(JParser.object());
  }
}

File: typescript/parser/src/parser/json/jparser.ts
import { Parser } from '../common/parser'
import { JValue } from './type/jvalue'
import * as P from '../common/parser'
import { JNullParser } from './jnull-parser';
import { JBoolParser } from './jbool-parser';
import { JStringParser } from './jstring-parser';
import { JNumberParser } from './jnumber-parser';
import { JArrayParser } from './jarray-parser';
import { JObjectParser } from './jobject-parser';

export interface JParser<T extends JValue> extends Parser<T> { }

export function values(): JParser<JValue>  {
	return P.or(P.or(P.or(P.or(P.or(new JNullParser(), new JBoolParser()), new JStringParser()), new JNumberParser()), new JArrayParser()), new JObjectParser())
}
